lib/vdc/DCUGRID.cpp

```diff
@@ -0,0 +1,627 @@
+
+        if (meshName.empty()) { continue; }
+
+        if (_uGridMeshMap.find(meshName) == _uGridMeshMap.end()) {
+            SetDiagMsg("No valid mesh for variable named %s", varName.c_str());
+            continue;
+        }
+
+        // Need to generate a unique mesh name for layered grids since
+        // UGRID allows 2D and 3D layered variables to share same mesh,
+        // but DC class does not.
+        //
+        vector<string> dimNames = ncdfc->GetSpatialDimNames(varName);
+        if (dimNames.size() == 2) { meshName += _getLayeredVerticalCoordVar(ncdfc, varName); }
+
+        // Only node-centered variables supported currently
+        //
+        string locationName = getAttString(ncdfc, varName, locationAttName);
+
+        if (locationName != "node") {
+            SetDiagMsg("Only node-centered data supported for variable named %s", varName.c_str());
+            continue;
+        }
+
+        string coordName;
+        bool   ok = _getVarTimeCoords(ncdfc, varName, coordName);
+        if (!ok) {
+            SetDiagMsg("Invalid variable : %s", varName.c_str());
+            continue;
+        }
+
+        string units;
+        ncdfc->GetAtt(varName, "units", units);
+
+        double mv;
+        bool   has_missing = ncdfc->GetMissingValue(varName, mv);
+
+        if (!has_missing) {
+            dataVarsMap[varName] = DataVar(varName, units, DC::FLOAT, periodic, meshName, coordName, DC::Mesh::NODE);
+        } else {
+            dataVarsMap[varName] = DataVar(varName, units, DC::FLOAT, periodic, meshName, coordName, DC::Mesh::NODE, mv);
+        }
+
+        int rc = DCUtils::CopyAtt(*ncdfc, varName, dataVarsMap[varName]);
+        if (rc < 0) return (-1);
+    }
+
+    return (0);
+}
+
+int DCUGRID::OpenVariableRead(size_t ts, string varname, int level, int lod)
+{
+    // Pass through normal open operation if this is not a synthesized
+    // connectivity variable
+    //
+    if (_faceNodeConnectivityMap.find(varname) == _faceNodeConnectivityMap.end()) { return (DC::OpenVariableRead(ts, varname, level, lod)); }
+
+    FileTable::FileObject *f = new FileTable::FileObject(ts, varname);
+    return (_fileTable.AddEntry(f));
+}
+
+int DCUGRID::Read(int fd, int *data)
+{
+    DC::FileTable::FileObject *f = _fileTable.GetEntry(fd);
+
+    if (!f) {
+        SetErrMsg("Invalid file descriptor : %d", fd);
+        return (-1);
+    }
+
+    // Pass through normal read operation if this is not a synthesized
+    // connectivity variable
+    //
+    if (_faceNodeConnectivityMap.find(f->GetVarname()) == _faceNodeConnectivityMap.end()) { return (DC::Read(fd, data)); }
+
+    const vector<int> &connVar = _faceNodeConnectivityMap[f->GetVarname()];
+    std::copy_n(connVar.data(), connVar.size(), data);
+    return (0);
+}
+
+int DCUGRID::ReadRegion(int fd, const vector<size_t> &min, const vector<size_t> &max, int *data)
+{
+    DC::FileTable::FileObject *f = _fileTable.GetEntry(fd);
```

Comment:
Does this pointed object need to be recycled (freed or closed)? If so, why not put it in a unique_ptr ?

Reply:
Because the memory referenced is used outside of the scope of this method.

Reply:
Ok, as long as it is recycled by someone at some point.